src/process_utility.c

```diff
@@ -1263,43 +1430,112 @@ process_index_end(Node *parsetree, CollectedCommand *cmd)
 	hcache = ts_hypertable_cache_pin();
 	ht = ts_hypertable_cache_get_entry_rv(hcache, stmt->relation);
 
-	if (NULL != ht)
+	if (NULL == ht)
 	{
-		CreateIndexInfo info = {
-			.stmt = stmt,
-		};
+		ts_cache_release(hcache);
+		return false;
+	}
 
-		switch (cmd->type)
+	switch (cmd->type)
+	{
+		case SCT_Simple:
+			info.obj = cmd->d.simple.address;
+			break;
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("hypertables do not support this operation: "
+							"parsetree %s, type %d",
+							nodeToString(parsetree),
+							cmd->type)));
+			break;
+	}
+
+	if (info.extended_options.multitransaction)
+	{
+		Relation index = relation_open(info.obj.objectId, AccessShareLock);
+		LockRelId index_lock_relid = index->rd_lockInfo.lockRelId;
+
+		relation_close(index, NoLock);
+
+		/*
+		 * Lock the index for the remainder of the command. Since we're using
+		 * multiple transactions for index creation, a regular
+		 * tranaction-level lock won't prevent the index from being
+		 * concurrently ALTERed or DELETEed. Instead, we grab a session level
+		 * lock on the index, which we'll release when the command is
+		 * finished. (This is the same strategy postgres uses in CREATE INDEX
+		 * CONCURRENTLY)
+		 */
+		LockRelationIdForSession(&index_lock_relid, AccessShareLock);
+
+		/*
+		 * mark the hypertable's index as invalid until all the chunk indexes
+		 * are created. This allows us to determine if the CREATE INDEX
+		 * completed successfully or  not
+		 */
+		ts_indexing_mark_as_invalid(info.obj.objectId);
+		CacheInvalidateRelcacheByRelid(ht->main_table_relid);
+		CacheInvalidateRelcacheByRelid(info.obj.objectId);
+
+		ts_cache_release(hcache);
+
+		info.mctx = CurrentMemoryContext;
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		/*
+		 * foreach_chunk needs an open tranaction, it will be closed in the
```

Comment:
I think it's actually better to change foreach_chunk to start and end a transaction around  find_inheritance_children if there is no valid transaction and then not have a transaction around process_chunk. Maybe this behavior should be controlled by a parameter to foreach_chunk.

Reply:
hmm, I'll try a commit like that, and see how it looks